File: lua/chkgrp-native-min-extended.lua
#!/usr/bin/env lua

-- This file is part of chkgrp that is a program that checks if a user is a
-- member of a group.
--
-- Author: Johannes Findeisen <you@hanez.org> - 2025
-- Homepage: https://git.xw3.org/hanez/chkgrp
-- License: Apache-2.0 (see LICENSE)

local MAX_NAME = 256

local function read_passwd_gid(username)
    local f = io.open("/etc/passwd", "r")
    if not f then return nil end

    for line in f:lines() do
        local fields = {}
        for field in string.gmatch(line, "([^:]+)") do
            table.insert(fields, field)
        end
        if fields[1] == username then
            f:close()
            return tonumber(fields[4]) -- GID is 4th field
        end
    end

    f:close()
    return nil
end

local function check_group_membership(username, groupname, user_gid)
    local f = io.open("/etc/group", "r")
    if not f then return nil end

    for line in f:lines() do
        local fields = {}
        for field in string.gmatch(line, "([^:]+)") do
            table.insert(fields, field)
        end
        if fields[1] == groupname then
            local group_gid = tonumber(fields[3])
            if group_gid == user_gid then
                f:close()
                return true
            end

            for member in string.gmatch(fields[4] or "", "[^,]+") do
                if member == username then
                    f:close()
                    return true
                end
            end

            f:close()
            return false
        end
    end

    f:close()
    return nil
end

local username = arg[1]
local groupname = arg[2]

if not username or not groupname then
    os.exit(2)
end

if #username > MAX_NAME or #groupname > MAX_NAME then
    os.exit(2)
end

local user_gid = read_passwd_gid(username)
if not user_gid then
    os.exit(2)
end

local result = check_group_membership(username, groupname, user_gid)
if result == nil then
    os.exit(2)
elseif result == true then
    print("Yes")
    os.exit(0)
else
    print("No")
    os.exit(1)
end

File: chkgrp-native-min-extended.c
/**
 * chkgrp-min-native-extended is a program that checks if a user is a member
 * of a group.
 *
 * Author: Johannes Findeisen <you@hanez.org> - 2025
 * Homepage: https://git.xw3.org/hanez/chkusr
 * License: Apache-2.0 (see LICENSE)
 */

#include "version.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINE 1024
#define MAX_NAME 256

int main(int argc, char *argv[])
{
    if (argc != 3)
        return 2;

    const char *username = argv[1];
    const char *groupname = argv[2];

    if (strlen(username) > MAX_NAME || strlen(groupname) > MAX_NAME)
        return 2;

    // Step 1: Find user in /etc/passwd to get their primary GID
    FILE *passwd = fopen("/etc/passwd", "r");
    if (!passwd)
        return 2;

    char line[MAX_LINE];
    int user_gid = -1;
    while (fgets(line, sizeof(line), passwd)) {
        char *u = strtok(line, ":");
        if (!u || strcmp(u, username) != 0)
            continue;

        strtok(NULL, ":"); // skip password
        strtok(NULL, ":"); // skip UID
        char *gid_str = strtok(NULL, ":");
        if (!gid_str)
            break;

        user_gid = atoi(gid_str);
        break;
    }
    fclose(passwd);
    if (user_gid < 0)
        return 2;

    // Step 2: Find group in /etc/group and check membership or GID match
    FILE *group = fopen("/etc/group", "r");
    if (!group)
        return 2;

    int found = 0;
    while (fgets(line, sizeof(line), group)) {
        char *grp = strtok(line, ":");
        if (!grp || strcmp(grp, groupname) != 0)
            continue;

        strtok(NULL, ":"); // skip password
        char *gid_str = strtok(NULL, ":");
        if (!gid_str)
            break;

        int group_gid = atoi(gid_str);
        if (group_gid == user_gid) {
            puts("Yes");
            fclose(group);
            return 0;
        }

        char *members = strtok(NULL, ":\n");
        if (!members)
            break;

        char *m = strtok(members, ",");
        while (m) {
            if (!strcmp(m, username)) {
                puts("Yes");
                fclose(group);
                return 0;
            }
            m = strtok(NULL, ",");
        }

        found = 1;
        break;
    }

    fclose(group);
    if (!found)
        return 2;

    puts("No");
    return 1;
}

